myjsonrpc.pm

```diff
@@ -95,22 +88,16 @@ sub read_json ($socket, $cmd_token = undef, $multi = undef) {
         }
 
         # wait for next read
-        my @res = $s->can_read;
-        while (!@res) {
+        until (my @res = $s->can_read) {
             # throw an error except can_read has been interrupted
             my $error = $!;
             confess "ERROR: unable to wait for JSON reply: $error\n" unless $!{EINTR};
             # try again if can_read's underlying system call has been interrupted as suggested by the perlipc documentation
-            bmwqemu::diag("read_json($fd): can_read's underlying system call has been interrupted, trying again\n") if DEBUG_JSON || $bmwqemu::vars{DEBUG_JSON_RPC};
-            @res = $s->can_read;
+            bmwqemu::diag("read_json($fd): can_read's underlying system call has been interrupted, trying again\n") if is_debug;
         }
 
         my $qbuffer;
-        my $bytes = sysread($socket, $qbuffer, READ_BUFFER);
-        if (!$bytes) {
-            bmwqemu::fctwarn("sysread failed: $!") if DEBUG_JSON || $bmwqemu::vars{DEBUG_JSON_RPC};
-            return;
-        }
+        my $bytes = sysread($socket, $qbuffer, READ_BUFFER) or do { bmwqemu::fctwarn("sysread failed: $!") if is_debug and return undef };
```

Comment:
This is really hard to read. It is really unclear what the left-hand-side of the `and` evaluates to if `is_debug` is false and it also looks like `is_debug and return undef` is all just part of the `if` condition although the `and return undef` part is actually not (meant) to be part of it.

I think the following will work (provided `bmwqemu::fctwarn` returns undef as well):
```suggestion
        my $bytes = sysread($socket, $qbuffer, READ_BUFFER) or return (is_debug ? bmwqemu::fctwarn("sysread failed: $!") : undef;
```

Reply:
But if we have a `do` block there's no reason to just use multiple statements separated by `;`.

Reply:
I don't have a problem with the original code, and putting it in one line is also just cheating with code coverage, and explicitly marking those lines as uncoverable would make it more clear that this **is** actually uncovered.